Added plugin system, added @git/fetch, @install/node, echo and @git/gitignore

FishingHacks opened the pull request.

FishingHacks commented:
#2 - ignore that, I thought it was on FishingHacks/redstart ^^

RedCrafter07 commented:
You may rename some variables, as they're prefixed with "_" which indicates they are not being used. Thanks :)

RedCrafter07 commented:
Also, have you pulled the current version before creating this request?

RedCrafter07 commented:
If you'd include timestamps for every module, it would be great. Thank you 👍

FishingHacks commented:
done

RedCrafter07 commented:
I can't see the timestamps anywhere 🤔

FishingHacks commented:
I removed them, cuz i think they're unnecessary until I've implemented top-level config

RedCrafter07 commented:
I think you should see how long a plugin took to do its work

RedCrafter07 commented:
But I can implement it myself then. Thanks for these great ideas 👍 😃

RedCrafter07 commented:
I'll look it over again in a few minutes and probably merge it if I don't find anything.

FishingHacks commented:
added a commit

RedCrafter07 commented:
For @git/fetch, will it be pulled? I think it would fail because of the config file... But I'm not sure.

FishingHacks commented:
Yes, it will be pulled. Why should it fail?

RedCrafter07 reviewed (changes requested):
It can't get the remote: (added code for logging the error)
Checking git...
error: No such remote 'origin'

This is a mistake as freshly initialized repositories don't have an origin remote.

FishingHacks commented:
See the problem, fixed in the next commit

RedCrafter07 commented:
And I'm not really sure about the cwd logic... The cwd should be grabbed from process.cwd() in my opinion

FishingHacks commented:
> And I'm not really sure about the cwd logic... The cwd should be grabbed from process.cwd() in my opinion

process.cwd() might not reflect the directory, the .rsproj-file is in. However, when someone puts a relative path into the .rsproj-file, they assume it will be relative to the directory, the .rsproj-file is in, not some arbitrary other.

FishingHacks commented:
> [Fixed @git/fetch, added usage, added contribute.md](https://github.com/RedCrafter07/redstart/pull/3/commits/f8ccef31ae80a8711989e8dd529d26312bd1b7a7)

FishingHacks commented:
Now it should be fixed

RedCrafter07 commented:
Maybe you have a configs/starter.rsproj file 🤔 
Have you implemented the option to override it in the config?

FishingHacks commented:
what

RedCrafter07 commented:
Before your changes, there was a workDir: option. Is it implemented here too?

FishingHacks commented:
no, because top-level (general) config doesn't work yet

RedCrafter07 commented:
What do you mean with "top-level config"?

RedCrafter07 reviewed (changes requested):
@git/fetch
fatal: protocol '"https' is not supported

FishingHacks commented:
smth like this:

```
...modules

workDir: ./ # Doesn't work
dbgprint: true # Doesn't work
[modulename]
key: value # does work
```

FishingHacks commented:
> @git/fetch fatal: protocol '"https' is not supported

> repository: "https://github.com/..."

is wrong, use

> repository: https://github.com/...

RedCrafter07 commented:
> > @git/fetch fatal: protocol '"https' is not supported
> 
> > repository: "https://github.com/..."
> 
> is wrong, use
> 
> > repository: https://github.com/...

I do so.

RedCrafter07 commented:
```rsproj
echo, @git/fetch

[echo]
message: Hello, World! This should be followed by nothing# When you see this, there's a problem going on :<
color: green

[@git/fetch]
repository: https://github.com/RedCrafter07/redstart.git
```

RedCrafter07 commented:
It's a parsing issue I addressed [here](https://github.com/RedCrafter07/redstart/commit/fe4cf7aba179c7dae3c671b43676628f1f372bd2)

FishingHacks commented:
```ts
function parseLine(line: string): null | { key: string; value: string } {
    if (line.includes("\n")) throw new Error("line is not a line");
    if (line.match(/^([^:]+): *[^\n\r]+$/)) {
        const split = line.split(":");
        const key = split.shift()?.trim();
        if (!key) return null;
        if (split.length < 1) return null;
        split[0] = split[0].trimStart();
        split[split.length - 1] = split[split.length - 1].trimEnd();
        return { key, value: split.join(":") };
    } else return null;
}
```

RedCrafter07 commented:
It doesn't work tho

RedCrafter07 commented:
> My configuration:
> 
> ```
> echo, @git/fetch
> 
> [echo]
> message: Hello, World! This should be followed by nothing# When you see this, there's a problem going on :<
> color: green
> 
> [@git/fetch]
> repository: https://github.com/RedCrafter07/redstart.git
> ```

Like seen here.

FishingHacks commented:
To test if it actually works, add

[echo]

message: This is a test :::

, it should print out This is a test :::, and it does that for me

RedCrafter07 commented:
> 

Works for me.

FishingHacks commented:
`git remote get-url origin`, output?

RedCrafter07 commented:
"https://github.com/RedCrafter07/redstart.git" although defined as seen in my configuration.

RedCrafter07 commented on src/modules/@git/fetch.ts:
```diff
@@ -0,0 +1,64 @@
+            if (remote.stdout.toString() !== config.repository) {
+                if (remote.stdout.toString() === "") {
+                    if (
+                        spawnSync(
+                            "git",
+                            [
+                                "remote",
+                                "add",
+                                "origin",
+                                `"${config.repository}"`,
+                            ],
+                            { cwd }
+                        ).status !== 0
```
This is the mistake.

FishingHacks replied:
oh, it is

RedCrafter07 replied:
You see, not my mistake 😁

FishingHacks replied:
I added that, cuz i forgor that it isn't executed like in the shell

RedCrafter07 replied:
Good that I reviewed it in detail 😅

FishingHacks commented:
should work now

RedCrafter07 commented:
You may use a [prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) like I do so you can help me with formatting in the future 👀 😅

RedCrafter07 commented:
Another error occurred while pulling the repo:
```
Checking git... From https://github.com/RedCrafter07/redstart
 * [new branch]      main       -> origin/main
 * [new tag]         v1.0.1     -> v1.0.1
 * [new tag]         v1.0.0     -> v1.0.0
There is no tracking information for the current branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> master


✖ Couldn't fetch remote repository`
```

RedCrafter07 commented on src/modules/@git/fetch.ts:
```diff
@@ -52,7 +52,7 @@ export default {
                     });
             }
-            if (spawnSync("git", ["pull"], { cwd }).status !== 0)
+            if (spawnSync("git", ["pull", "--set-upstream-to=origin/" + (config.branch || "master")], { cwd }).status !== 0)
```
When executing:
```
error: unknown option `set-upstream-to=origin/main'
usage: git pull [<options>] [<repository> [<refspec>...]]
...
```

FishingHacks commented:
fixed

RedCrafter07 commented:
[!] Branch main not found!

RedCrafter07 commented:
I'm still getting the error 😞

FishingHacks commented:
ik

FishingHacks commented:
btw, you should use yaml as the highlighter for .rsproj files. only the @ in [@a/b] is not hightlighted

RedCrafter07 reviewed (changes requested):
Branches on remote are failing (added for clearance)

RedCrafter07 commented:
> btw, you should use yaml as the highlighter for .rsproj files. only the @ in [@a/b] is not hightlighted

Oh, true

FishingHacks commented:
bruh, now it doesn't even add the remote

FishingHacks commented:
häääää

FishingHacks commented:
console.log("'git checkout " + config.branch + "'");
spawnSync("git", ["checkout", config.branch])

when i run the command, that gets logged to the console, everything works correctly

FishingHacks commented:
y

FishingHacks commented:
lol

RedCrafter07 commented:
I'm sure you'll figure it out 👏

RedCrafter07 commented:
Errors with "Couldn't fetch remote repository"

FishingHacks commented:
> [Added @test/js, @git/fetch now really works](https://github.com/RedCrafter07/redstart/pull/3/commits/d52a47c2b9639340f67e0ca2e014aca6dffbe78a)

FishingHacks reviewed (commented):
Everything works as expected!

RedCrafter07 commented:
It works. Nice

RedCrafter07 commented:
But you logged a git command

RedCrafter07 commented:
Fixed.

FishingHacks commented:
thx
